src/main/java/iterator/RichIterator.java

```diff
@@ -88,30 +136,40 @@ default boolean contains(A elem) {
      * @return the collection
      */
     default <C extends Collection<A>> C toCollection(Supplier<C> collectionFactory) {
-        throw new NotImplementedException();
+        C collection = collectionFactory.get();
+        while (hasNext()) {
+            collection.add(next());
+        }
+
+        return collection;
     }
 
     /**
      * @return a list built from the iterator's elements
      */
     default List<A> toList() {
-        throw new NotImplementedException();
+        return toCollection(ArrayList::new);
```

Comment:
```suggestion
        return toCollection(LinkedList::new);
```

Array list is not optimal for this because the size of the iterator is unknown. `LinkedList` provides a better time solution because adding an element to it is always O(1) but in `ArrayList` it is not always the case.